python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -944,41 +972,119 @@ def _get_entities_from_payload(self, data):
 
             for entity in data["entity_ids"]:
                 # Did we get an entity list, or a list of entity ids?
                 if isinstance(entity, dict):
-                    if "project" not in entity:
-                        entity["project"] = project_entity
+                    # If we were passed a usable project entity from the web app, we
+                    # can trust that and add it to our entity. If we didn't, then we'll
+                    # have to query it.
+                    if entity.get("project") is None:
+                        if project_entity is None:
+                            entity["project"] = self._get_entity_parent_project(entity)
+                        else:
+                            entity["project"] = project_entity
 
                     entities.append(entity)
                 else:
-                    entities.append(
-                        dict(
-                            type=data["entity_type"],
-                            id=entity,
-                            project=project_entity,
-                        )
+                    entity = dict(
+                        type=data["entity_type"],
+                        id=entity,
                     )
+                    # If we were passed a usable project entity from the web app, we
+                    # can trust that and add it to our entity. If we didn't, then we'll
+                    # have to query it.
+                    if project_entity is None:
+                        entity["project"] = project_entity
+                    else:
+                        entity["project"] = self._get_entity_parent_project(entity)
+
+                    entities.append(entity)
+
+            # If we were not given a usable project entity, we can pull it from an
+            # entity we've just extracted. This doesn't cover the case of receiving
+            # multiple entities from the web app from different projects, but this isn't
+            # the only place where we're going to suffer there.
+            if project_entity is None:
+                project_entity = entities[0]["project"]
+                if project_entity is None:
+                    raise RuntimeError("Unable to determine a project entity from data: %s" % data)
+
             return (project_entity, entities)
         else:
             raise RuntimeError("Unable to determine an entity from data: %s" % data)
 
     @sgtk.LogManager.log_timing
+    def _get_entity_parent_project(self, entity):
+        """
+        Gets the project entity that the given entity is linked to.
+
+        :param dict entity: A standard Shotgun entity dictionary.
+
+        :returns: A standard Shotgun Project entity.
+        :rtype: dict
+        """
+        if entity.get("project") is not None:
+            return entity["project"]
+
+        if entity["type"] == "Project":
+            return entity
+
+        if self.ENTITY_PARENT_PROJECTS not in self._cache:
```

Comment:
Have you added this cache because you perceived a performance problem or because you thought it could turn into one someday? An entity move from one project to another would not be detected when getting the parent project because of this cache. Not a big concern, I know, but if performance is not an issue at the moment, maybe its safer to not add a cache for now, while also add a comment about potentially adding one in the future and mention the dangers of doing so.

Reply:
With that being said, the pattern of "if missing key/value then insert" can be written as a one liner as

```
project_cache = self._cache.setdefault(self.ENTITY_PARENT_PROJECTS, {})
if entity["id"]  not in project_cache:
```

Reply:
If also saves the code from having to reevaluate `self._cache[self.ENTITY_PARENT_PROJECTS]`.